rts/python/reader.py

```diff
@@ -257,8 +264,263 @@ def verify_array_dims(l, dims):
+    }
+
+def read_bin_read_type_enum(f):
+    read_binname = get_chars(f, 4)
+
+    for (k,v) in FUTHARK_PRIMTYPES.items():
+        if v['binname'] == read_binname:
+            return k
+    panic(1, "binary-input: Did not recognize the type '%s'.\n", read_binname)
+
+def read_bin_ensure_scalar(f, expected_type):
+  dims = read_bin_i8(f)
+
+  if bin_dims != 0:
+      panic(1, "binary-input: Expected scalar (0 dimensions), but got array with %i dimensions.\n", bin_dims)
+
+  bin_type_enum = read_bin_read_type_enum(f)
+  if bin_type_enum != expected_type:
+      panic(1, "binary-input: Expected scalar of type %s but got scalar of type %s.\n",
+            FUTHARK_PRIMTYPES[expected_type]['type_name'],
+            FUTHARK_PRIMTYPES[bin_type_enum]['type_name'])
+
+# ------------------------------------------------------------------------------
+# General interface for reading Primitive Futhark Values
+# ------------------------------------------------------------------------------
+
+def read_general(f, ty):
+    if read_is_binary(f):
+        read_bin_ensure_scalar(ty)
+        return FUTHARK_PRIMTYPES[ty]['bin_reader'](f)
+    return FUTHARK_PRIMTYPES[ty]['str_reader'](f)
+
+def read_i8(f):
+    return read_general(f, FUTHARK_INT8)
+
+def read_i16(f):
+    return read_general(f, FUTHARK_INT16)
+
+def read_i32(f):
+    return read_general(f, FUTHARK_INT32)
+
+def read_i64(f):
+    return read_general(f, FUTHARK_INT64)
+
+def read_u8(f):
+    return read_general(f, FUTHARK_UINT8)
+
+def read_u16(f):
+    return read_general(f, FUTHARK_UINT16)
+
+def read_u32(f):
+    return read_general(f, FUTHARK_UINT32)
+
+def read_u64(f):
+    return read_general(f, FUTHARK_UINT64)
+
+def read_f32(f):
+    return read_general(f, FUTHARK_FLOAT32)
+
+def read_f64(f):
+    return read_general(f, FUTHARK_FLOAT64)
+
+def read_bool(f):
+    return read_general(f, FUTHARK_BOOL)
+
+def read_array(f, expected_type, type_name, rank, ctype):
+    if not read_is_binary(f):
+        str_reader = FUTHARK_PRIMTYPES[expected_type]['str_reader']
+        return read_str_array(f, str_reader, type_name, rank, ctype)
+
+    bin_rank = read_bin_u8(f)
+
+    if bin_rank != rank:
+        panic(1, "binary-input: Expected %i dimensions, but got array with %i dimensions.\n",
+              rank, bin_rank)
+
+    bin_type_enum = read_bin_read_type_enum(f)
+    if expected_type != bin_type_enum:
+        panic(1, "binary-input: Expected %iD-array with element type '%s' but got %iD-array with element type '%s'.\n",
+              rank, FUTHARK_PRIMTYPES[expected_type]['type_name'],
+              rank, FUTHARK_PRIMTYPES[bin_type_enum]['type_name'])
+
+    shape = []
+    elem_count = 1
+    for i in range(rank):
+        bin_size = read_bin_u64(f)
+        elem_count *= bin_size
+        shape.append(bin_size)
+
+    bin_reader = FUTHARK_PRIMTYPES[bin_type_enum]['bin_reader']
+
+    arr = np.empty(elem_count, dtype=ctype)
+    for i in range(elem_count):
+        arr[i] = bin_reader(f)
```

Comment:
I think it could be faster by unpacking the whole array in one pass (either `struct.unpack(fmt*array_size, input)` or using `np.fromfile(f, dtype, array_size)`